velox/exec/tests/RowNumberFuzzer.cpp

```diff
@@ -0,0 +1,541 @@
+    const std::vector<std::string>& keyNames,
+    const std::vector<TypePtr>& keyTypes) {
+  std::vector<std::string> names = keyNames;
+  std::vector<TypePtr> types = keyTypes;
+  // Add up to 3 payload columns.
+  const auto numPayload = randInt(0, 3);
+  for (auto i = 0; i < numPayload; ++i) {
+    names.push_back(fmt::format("c{}", i + keyNames.size()));
+    types.push_back(vectorFuzzer_.randType(2 /*maxDepth*/));
+  }
+
+  const auto inputType = ROW(std::move(names), std::move(types));
+  std::vector<RowVectorPtr> input;
+  input.reserve(FLAGS_num_batches);
+  for (auto i = 0; i < FLAGS_num_batches; ++i) {
+    input.push_back(vectorFuzzer_.fuzzInputRow(inputType));
+  }
+
+  return input;
+}
+
+RowNumberFuzzer::PlanWithSplits RowNumberFuzzer::makeDefaultPlan(
+    const std::vector<std::string>& partitionKeys,
+    const std::vector<RowVectorPtr>& input) {
+  auto planNodeIdGenerator = std::make_shared<core::PlanNodeIdGenerator>();
+  std::vector<std::string> projectFields = partitionKeys;
+  projectFields.emplace_back("row_number");
+  auto plan = PlanBuilder()
+                  .values(input)
+                  .rowNumber(partitionKeys)
+                  .project(projectFields)
+                  .planNode();
+  return PlanWithSplits{plan};
+}
+
+RowVectorPtr RowNumberFuzzer::execute(
+    const PlanWithSplits& plan,
+    bool injectSpill) {
+  LOG(INFO) << "Executing query plan: " << plan.plan->toString(true, true);
+
+  AssertQueryBuilder builder(plan.plan);
+  if (!plan.splits.empty()) {
+    builder.splits(plan.splits);
+  }
+
+  std::shared_ptr<TempDirectoryPath> spillDirectory;
+  int32_t spillPct{0};
+  if (injectSpill) {
+    spillDirectory = exec::test::TempDirectoryPath::create();
+    builder.config(core::QueryConfig::kSpillEnabled, true)
+        .config(core::QueryConfig::kRowNumberSpillEnabled, true)
+        .spillDirectory(spillDirectory->getPath());
+    spillPct = 10;
+  }
+
+  ScopedOOMInjector oomInjector(
+      []() -> bool { return folly::Random::oneIn(10); },
+      10); // Check the condition every 10 ms.
+  if (FLAGS_enable_oom_injection) {
+    oomInjector.enable();
+  }
+
+  TestScopedSpillInjection scopedSpillInjection(spillPct);
+  RowVectorPtr result;
+  try {
+    result = builder.copyResults(pool_.get());
+  } catch (VeloxRuntimeError& e) {
+    if (FLAGS_enable_oom_injection &&
+        e.errorCode() == facebook::velox::error_code::kMemCapExceeded &&
+        e.message() == ScopedOOMInjector::kErrorMessage) {
+      // If we enabled OOM injection we expect the exception thrown by the
+      // ScopedOOMInjector.
+      return nullptr;
+    }
+
+    throw e;
+  }
+
+  if (VLOG_IS_ON(1)) {
+    VLOG(1) << std::endl << result->toString(0, result->size());
+  }
+  // Wait for the task to be destroyed before start next query execution to
+  // avoid the potential interference of the background activities across query
+  // executions.
+  waitForAllTasksToBeDeleted();
```

Comment:
Shall we put waitForAllTasksToBeDeleted() in folly::makeGuard to call that in case of exception as well? Thanks!